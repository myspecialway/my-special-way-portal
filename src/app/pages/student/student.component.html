<div data-test-id="students-page" class="main-content msw-student-container">
  <div class="container-fluid">
    <div class="header-menu-container msw-table-tools-header">
      <div class="header-table-count">
        <div>
          <span class="header-table-count-label">תלמידים</span>
          <span class="header-table-count-number">({{dataSource.data.length}})</span>
        </div>
      </div>
      <div class="example-header msw-table-tools-header">
        <button data-test-id="add-new-student-button" mat-raised-button color="primary" routerLink="/student/_new_"  *ngIf="mayAddStudent">
          <i class="material-icons">
            add
          </i>
          הוסף תלמיד חדש
        </button>
      </div>
    </div>
    <div class="row">
      <div class="col-lg-12 col-md-12 col-sm-12">
        <mat-table [dataSource]="dataSource" class="msw-table" matSort matSortDisableClear>
          <ng-container matColumnDef="studentName">
            <mat-header-cell *matHeaderCellDef>שם</mat-header-cell>
            <mat-cell *matCellDef="let row" [routerLink]="['/student', row._id]">{{ row.firstname }} {{ row.lastname }}</mat-cell>
          </ng-container>

          <ng-container matColumnDef="gradeId">
            <mat-header-cell *matHeaderCellDef>כיתה</mat-header-cell>
            <mat-cell *matCellDef="let row" [routerLink]="['/student', row._id]">{{(row.class && row.class.name) || '' }}</mat-cell>
          </ng-container>

          <ng-container matColumnDef="username">
            <mat-header-cell *matHeaderCellDef mat-sort-header>שם משתמש</mat-header-cell>
            <mat-cell *matCellDef="let row" class="username" [routerLink]="['/student', row._id]">{{ row.username }}</mat-cell>
          </ng-container>

          <ng-container matColumnDef="deleteUser">
            <mat-header-cell *matHeaderCellDef></mat-header-cell>
            <mat-cell *matCellDef="let row">
              <button mat-icon-button color="primary" [attr.data-test-id]="'delete-user-button-' + row.username" [disabled]="!mayDeleteStudent"
                      (click)="deleteStudent(row._id, row.firstname, row.lastname, row.gradeId, row.gender)"><i class="material-icons"> delete </i> </button>
            </mat-cell>
          </ng-container>

          <mat-header-row *matHeaderRowDef="displayedColumns"></mat-header-row>
          <mat-row *matRowDef="let row; columns: displayedColumns;"></mat-row>
        </mat-table>
      </div>
    </div>
  </div>
</div>
