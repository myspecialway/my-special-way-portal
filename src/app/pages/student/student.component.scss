@import '../../../assets/scss/msw-table';

.header-table-count {
  width: 164px;
  height: 38px;
  font-size: 32px;
  font-weight: normal;
  font-style: normal;
  font-stretch: normal;
  line-height: normal;
  letter-spacing: 0.6px;
  text-align: right;
  color: #222222;
}
.header-table-count-number {
  font-size: 18px;
}

.header-table-count-label {
  padding-left: 10px;
}

.header-menu-container {
  padding-bottom: 29px;
}

.msw-student-container .mat-cell:last-child,
.msw-student-container .mat-header-cell:last-child {
  flex-grow: 0;
  flex-basis: 100px;
}
.msw-student-container .mat-header-cell {
  font-weight: bold;
}

.mat-column-deleteUser button {
  display: none;
}
.mat-row:hover {
  .mat-cell {
    color: #8c37cd;
  }
  .mat-column-deleteUser button {
    display: block;
  }
}
