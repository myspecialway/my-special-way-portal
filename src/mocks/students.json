[
  {
    "_id": "5b0affc6d9c920b8ba7e64bf",
    "userName": "ullamco",
    "password": "id",
    "firstName": "Pearl",
    "lastName": "Moss",
    "gradeId": "5b0116f814169898640db32f"
  },
  {
    "_id": "5b0affc6530d72659ead28a7",
    "userName": "et",
    "password": "dolore",
    "firstName": "Autumn",
    "lastName": "Dodson",
    "gradeId": "5b0116f814169898640db32f"
  },
  {
    "_id": "5b0affc6ed0b8878f6cc6b21",
    "userName": "deserunt",
    "password": "velit",
    "firstName": "Wyatt",
    "lastName": "Lott",
    "gradeId": "5b0116f8b9e91d54d315372c"
  },
  {
    "_id": "5b0affc642a84150497bb283",
    "userName": "nulla",
    "password": "cupidatat",
    "firstName": "Small",
    "lastName": "Phillips",
    "gradeId": "5b0116f8b9e91d54d315372c"
  },
  {
    "_id": "5b0affc6f30e239b877a8c13",
    "userName": "magna",
    "password": "consectetur",
    "firstName": "Maura",
    "lastName": "Doyle",
    "gradeId": "5b0116f814169898640db32f"
  },
  {
    "_id": "5b0affc624225ed42728ebc3",
    "userName": "culpa",
    "password": "amet",
    "firstName": "Hendricks",
    "lastName": "Alvarado",
    "gradeId": "5b0116f896b61b3a5c135a1d"
  },
  {
    "_id": "5b0affc6c994aba7ad4bfa50",
    "userName": "officia",
    "password": "veniam",
    "firstName": "Maricela",
    "lastName": "Austin",
    "gradeId": "5b0116f870a67d3a47e9f235"
  },
  {
    "_id": "5b0affc611aec69a7e96cca1",
    "userName": "dolore",
    "password": "nostrud",
    "firstName": "Cooley",
    "lastName": "Summers",
    "gradeId": "5b0116f814169898640db32f"
  },
  {
    "_id": "5b0affc6d44ce84b0f2220ca",
    "userName": "elit",
    "password": "deserunt",
    "firstName": "Avila",
    "lastName": "Hendrix",
    "gradeId": "5b0116f841bc308bff51f0b8"
  },
  {
    "_id": "5b0affc6870467d4b7ec36d2",
    "userName": "velit",
    "password": "aliquip",
    "firstName": "Craig",
    "lastName": "Orr",
    "gradeId": "5b0116f846bd827beedec579"
  },
  {
    "_id": "5b0affc645e2481e505dbd31",
    "userName": "ut",
    "password": "ex",
    "firstName": "Colleen",
    "lastName": "Martinez",
    "gradeId": "5b0116f82074861f963642f3"
  },
  {
    "_id": "5b0affc6c8aead3065a32dbd",
    "userName": "in",
    "password": "duis",
    "firstName": "Lavonne",
    "lastName": "West",
    "gradeId": "5b0116f814169898640db32f"
  },
  {
    "_id": "5b0affc618e7ce7bf79bb5ed",
    "userName": "ad",
    "password": "qui",
    "firstName": "Melton",
    "lastName": "Day",
    "gradeId": "5b0116f87554b3534cfde9b7"
  },
  {
    "_id": "5b0affc67e2e1e9882e1346b",
    "userName": "cillum",
    "password": "aliqua",
    "firstName": "Fisher",
    "lastName": "Franks",
    "gradeId": "5b0116f896b61b3a5c135a1d"
  },
  {
    "_id": "5b0affc68ca6af1d03c61254",
    "userName": "ea",
    "password": "qui",
    "firstName": "Dawson",
    "lastName": "Collier",
    "gradeId": "5b0116f855ecf6fd19ae77fa"
  },
  {
    "_id": "5b0affc6fe71ef20902a690f",
    "userName": "veniam",
    "password": "commodo",
    "firstName": "Leila",
    "lastName": "Daniels",
    "gradeId": "5b0116f8b9e91d54d315372c"
  },
  {
    "_id": "5b0affc62c8b9e6eb1c58f25",
    "userName": "ut",
    "password": "amet",
    "firstName": "Janet",
    "lastName": "Bonner",
    "gradeId": "5b0116f832a6de313076c3b1"
  },
  {
    "_id": "5b0affc67710d211f61084bb",
    "userName": "cupidatat",
    "password": "ut",
    "firstName": "Mildred",
    "lastName": "Henry",
    "gradeId": "5b0116f82074861f963642f3"
  },
  {
    "_id": "5b0affc6df4b3abe96e559be",
    "userName": "ullamco",
    "password": "esse",
    "firstName": "Krista",
    "lastName": "Montgomery",
    "gradeId": "5b0116f8cc498da37f361b2f"
  },
  {
    "_id": "5b0affc68a927ba472e9e8ea",
    "userName": "magna",
    "password": "commodo",
    "firstName": "Cleveland",
    "lastName": "Jensen",
    "gradeId": "5b0116f855ecf6fd19ae77fa"
  },
  {
    "_id": "5b0affc6f21dd40b536419bf",
    "userName": "aliquip",
    "password": "nisi",
    "firstName": "Alyson",
    "lastName": "Hayes",
    "gradeId": "5b0116f832a6de313076c3b1"
  },
  {
    "_id": "5b0affc6751da69d5df24437",
    "userName": "fugiat",
    "password": "voluptate",
    "firstName": "Robin",
    "lastName": "Shepard",
    "gradeId": "5b0116f8b9e91d54d315372c"
  },
  {
    "_id": "5b0affc6a94a1f487b2c96a3",
    "userName": "aute",
    "password": "aliqua",
    "firstName": "Crawford",
    "lastName": "Flores",
    "gradeId": "5b0116f8cc498da37f361b2f"
  },
  {
    "_id": "5b0affc6312031d4789147de",
    "userName": "occaecat",
    "password": "voluptate",
    "firstName": "Chan",
    "lastName": "Rodriquez",
    "gradeId": "5b0116f870788b1e9f99b814"
  },
  {
    "_id": "5b0affc62b5a3a4428f10149",
    "userName": "proident",
    "password": "magna",
    "firstName": "Leach",
    "lastName": "Padilla",
    "gradeId": "5b0116f8895af89de5563236"
  },
  {
    "_id": "5b0affc6aa154d39c94fd0a3",
    "userName": "minim",
    "password": "in",
    "firstName": "Garrison",
    "lastName": "Delaney",
    "gradeId": "5b0116f832a6de313076c3b1"
  },
  {
    "_id": "5b0affc6622c30bcd643c79b",
    "userName": "aliquip",
    "password": "ad",
    "firstName": "Amanda",
    "lastName": "Mcintyre",
    "gradeId": "5b0116f8897ec61406fc591f"
  },
  {
    "_id": "5b0affc6402ac5cd2c8405b3",
    "userName": "laborum",
    "password": "reprehenderit",
    "firstName": "Nola",
    "lastName": "Hardy",
    "gradeId": "5b0116f8cc498da37f361b2f"
  },
  {
    "_id": "5b0affc696ecac04e9cf4c5d",
    "userName": "duis",
    "password": "ex",
    "firstName": "Ochoa",
    "lastName": "Mckinney",
    "gradeId": "5b0116f8895af89de5563236"
  },
  {
    "_id": "5b0affc6c9e146b966f04fe7",
    "userName": "aute",
    "password": "commodo",
    "firstName": "Annie",
    "lastName": "Huffman",
    "gradeId": "5b0116f814169898640db32f"
  },
  {
    "_id": "5b0affc689f6eb7dc6a224d9",
    "userName": "irure",
    "password": "incididunt",
    "firstName": "Gould",
    "lastName": "Cook",
    "gradeId": "5b0116f814169898640db32f"
  },
  {
    "_id": "5b0affc60a941198ef215578",
    "userName": "enim",
    "password": "in",
    "firstName": "Little",
    "lastName": "Castaneda",
    "gradeId": "5b0116f870a67d3a47e9f235"
  },
  {
    "_id": "5b0affc6850bcc1e1309e252",
    "userName": "Lorem",
    "password": "in",
    "firstName": "Glenda",
    "lastName": "Morrison",
    "gradeId": "5b0116f8cc498da37f361b2f"
  },
  {
    "_id": "5b0affc6d42bb374c8a99bb7",
    "userName": "occaecat",
    "password": "qui",
    "firstName": "Randall",
    "lastName": "Griffin",
    "gradeId": "5b0116f896b61b3a5c135a1d"
  },
  {
    "_id": "5b0affc68654d3556b0ffcbf",
    "userName": "ad",
    "password": "sit",
    "firstName": "Alisha",
    "lastName": "Ferguson",
    "gradeId": "5b0116f8c1d99ef1b8e73245"
  },
  {
    "_id": "5b0affc6803d4919e3f9945c",
    "userName": "fugiat",
    "password": "nostrud",
    "firstName": "Ingrid",
    "lastName": "Banks",
    "gradeId": "5b0116f814169898640db32f"
  },
  {
    "_id": "5b0affc606f6a3dadb28e77e",
    "userName": "commodo",
    "password": "nostrud",
    "firstName": "Mullen",
    "lastName": "Watts",
    "gradeId": "5b0116f82074861f963642f3"
  },
  {
    "_id": "5b0affc6f2257593b792f0f0",
    "userName": "mollit",
    "password": "veniam",
    "firstName": "Marie",
    "lastName": "Hamilton",
    "gradeId": "5b0116f846bd827beedec579"
  },
  {
    "_id": "5b0affc6e53a624a962479f9",
    "userName": "fugiat",
    "password": "non",
    "firstName": "Sheree",
    "lastName": "Riley",
    "gradeId": "5b0116f867a266b544055fc3"
  },
  {
    "_id": "5b0affc62a13fb52908439ae",
    "userName": "est",
    "password": "cillum",
    "firstName": "Jacqueline",
    "lastName": "Craft",
    "gradeId": "5b0116f814169898640db32f"
  },
  {
    "_id": "5b0affc6839969093dcbe25f",
    "userName": "reprehenderit",
    "password": "exercitation",
    "firstName": "Chandra",
    "lastName": "Cabrera",
    "gradeId": "5b0116f8f0319d311adbd5b8"
  },
  {
    "_id": "5b0affc6b7f13e1e0799bad2",
    "userName": "deserunt",
    "password": "voluptate",
    "firstName": "Griffin",
    "lastName": "Cooley",
    "gradeId": "5b0116f8f0319d311adbd5b8"
  },
  {
    "_id": "5b0affc6d95c13a7d9c7cb85",
    "userName": "proident",
    "password": "laboris",
    "firstName": "Tasha",
    "lastName": "Herring",
    "gradeId": "5b0116f814169898640db32f"
  },
  {
    "_id": "5b0affc6e31a80faaf00afa9",
    "userName": "non",
    "password": "voluptate",
    "firstName": "Nikki",
    "lastName": "Rose",
    "gradeId": "5b0116f8cc498da37f361b2f"
  },
  {
    "_id": "5b0affc6f913b7e5ded15fec",
    "userName": "esse",
    "password": "ut",
    "firstName": "Lowery",
    "lastName": "Sutton",
    "gradeId": "5b0116f8ab446a3271359472"
  },
  {
    "_id": "5b0affc6e84807e38eb6e397",
    "userName": "aute",
    "password": "ipsum",
    "firstName": "Weber",
    "lastName": "Glass",
    "gradeId": "5b0116f867a266b544055fc3"
  },
  {
    "_id": "5b0affc68cd61b8f00fc9a77",
    "userName": "veniam",
    "password": "nisi",
    "firstName": "Delia",
    "lastName": "Marquez",
    "gradeId": "5b0116f870788b1e9f99b814"
  },
  {
    "_id": "5b0affc67c7311edd9cb3167",
    "userName": "reprehenderit",
    "password": "culpa",
    "firstName": "Maxine",
    "lastName": "Mueller",
    "gradeId": "5b0116f832a6de313076c3b1"
  },
  {
    "_id": "5b0affc627c93fdca8f3072b",
    "userName": "et",
    "password": "officia",
    "firstName": "Bright",
    "lastName": "Owen",
    "gradeId": "5b0116f896b61b3a5c135a1d"
  },
  {
    "_id": "5b0affc6d655a06e37d1e838",
    "userName": "quis",
    "password": "adipisicing",
    "firstName": "Macias",
    "lastName": "Norris",
    "gradeId": "5b0116f832a6de313076c3b1"
  },
  {
    "_id": "5b0affc6cbca08ac4a7e637a",
    "userName": "occaecat",
    "password": "incididunt",
    "firstName": "Knapp",
    "lastName": "Knox",
    "gradeId": "5b0116f855ecf6fd19ae77fa"
  },
  {
    "_id": "5b0affc65631ef5dc5c7f82e",
    "userName": "eu",
    "password": "elit",
    "firstName": "Kent",
    "lastName": "Ford",
    "gradeId": "5b0116f867a266b544055fc3"
  },
  {
    "_id": "5b0affc681efd88faca2970a",
    "userName": "dolore",
    "password": "consectetur",
    "firstName": "Duke",
    "lastName": "Shepherd",
    "gradeId": "5b0116f8cc498da37f361b2f"
  },
  {
    "_id": "5b0affc6ed8c74adc710d11c",
    "userName": "cupidatat",
    "password": "anim",
    "firstName": "Betsy",
    "lastName": "Simon",
    "gradeId": "5b0116f867a266b544055fc3"
  },
  {
    "_id": "5b0affc6c2502a738995aae7",
    "userName": "consectetur",
    "password": "incididunt",
    "firstName": "Dollie",
    "lastName": "Cobb",
    "gradeId": "5b0116f870788b1e9f99b814"
  },
  {
    "_id": "5b0affc6718c7b6934608881",
    "userName": "magna",
    "password": "officia",
    "firstName": "Richards",
    "lastName": "Roy",
    "gradeId": "5b0116f8895af89de5563236"
  },
  {
    "_id": "5b0affc6c9861698b6ede869",
    "userName": "anim",
    "password": "mollit",
    "firstName": "James",
    "lastName": "Donovan",
    "gradeId": "5b0116f896b61b3a5c135a1d"
  },
  {
    "_id": "5b0affc686645565978d367b",
    "userName": "sunt",
    "password": "officia",
    "firstName": "Gutierrez",
    "lastName": "Wilkerson",
    "gradeId": "5b0116f8ab446a3271359472"
  },
  {
    "_id": "5b0affc6b900bd9dd629a262",
    "userName": "in",
    "password": "est",
    "firstName": "Renee",
    "lastName": "Munoz",
    "gradeId": "5b0116f8895af89de5563236"
  },
  {
    "_id": "5b0affc61cab98ad9656a01c",
    "userName": "culpa",
    "password": "et",
    "firstName": "Jordan",
    "lastName": "Spence",
    "gradeId": "5b0116f8ab446a3271359472"
  },
  {
    "_id": "5b0affc624d37bbabd5a2a6a",
    "userName": "excepteur",
    "password": "non",
    "firstName": "Eunice",
    "lastName": "Wheeler",
    "gradeId": "5b0116f814169898640db32f"
  },
  {
    "_id": "5b0affc6bc91ba8ae64ab00b",
    "userName": "est",
    "password": "ut",
    "firstName": "Drake",
    "lastName": "Dennis",
    "gradeId": "5b0116f87554b3534cfde9b7"
  },
  {
    "_id": "5b0affc6c70f46ee8d22dc61",
    "userName": "eiusmod",
    "password": "eiusmod",
    "firstName": "Barbara",
    "lastName": "Short",
    "gradeId": "5b0116f832a6de313076c3b1"
  },
  {
    "_id": "5b0affc6a35032c839e2656b",
    "userName": "laboris",
    "password": "do",
    "firstName": "Stafford",
    "lastName": "Noble",
    "gradeId": "5b0116f8c1d99ef1b8e73245"
  },
  {
    "_id": "5b0affc65a90e96dd6d562e7",
    "userName": "aute",
    "password": "quis",
    "firstName": "Hanson",
    "lastName": "Potts",
    "gradeId": "5b0116f832a6de313076c3b1"
  },
  {
    "_id": "5b0affc6dc54894b77a66ce2",
    "userName": "elit",
    "password": "ex",
    "firstName": "Guerra",
    "lastName": "Gaines",
    "gradeId": "5b0116f814169898640db32f"
  },
  {
    "_id": "5b0affc6066723006fee501f",
    "userName": "magna",
    "password": "cillum",
    "firstName": "Figueroa",
    "lastName": "Ferrell",
    "gradeId": "5b0116f8895af89de5563236"
  },
  {
    "_id": "5b0affc64ae1143858545ce0",
    "userName": "amet",
    "password": "eiusmod",
    "firstName": "Becky",
    "lastName": "Hopper",
    "gradeId": "5b0116f8b9e91d54d315372c"
  },
  {
    "_id": "5b0affc6564e6ef7d3a7786e",
    "userName": "cillum",
    "password": "enim",
    "firstName": "Montoya",
    "lastName": "Patel",
    "gradeId": "5b0116f814169898640db32f"
  },
  {
    "_id": "5b0affc6cdaff675233e3266",
    "userName": "veniam",
    "password": "officia",
    "firstName": "Duncan",
    "lastName": "Jones",
    "gradeId": "5b0116f832a6de313076c3b1"
  },
  {
    "_id": "5b0affc6ceb2e3faf8d8d6a2",
    "userName": "incididunt",
    "password": "do",
    "firstName": "Norma",
    "lastName": "Merritt",
    "gradeId": "5b0116f8b9e91d54d315372c"
  },
  {
    "_id": "5b0affc61489b61ffe866c9d",
    "userName": "labore",
    "password": "ullamco",
    "firstName": "Whitley",
    "lastName": "Alston",
    "gradeId": "5b0116f867a266b544055fc3"
  },
  {
    "_id": "5b0affc6ae19cf32ae0c2b92",
    "userName": "ut",
    "password": "nulla",
    "firstName": "Brooks",
    "lastName": "Eaton",
    "gradeId": "5b0116f846bd827beedec579"
  },
  {
    "_id": "5b0affc6dfe3a5b65b9e3ce8",
    "userName": "nostrud",
    "password": "esse",
    "firstName": "Marcella",
    "lastName": "Morris",
    "gradeId": "5b0116f896b61b3a5c135a1d"
  },
  {
    "_id": "5b0affc6cdd0bfd26d95c216",
    "userName": "enim",
    "password": "sunt",
    "firstName": "Espinoza",
    "lastName": "England",
    "gradeId": "5b0116f867a266b544055fc3"
  },
  {
    "_id": "5b0affc63b3222a9d278d32d",
    "userName": "sunt",
    "password": "voluptate",
    "firstName": "Roslyn",
    "lastName": "Tucker",
    "gradeId": "5b0116f8f0319d311adbd5b8"
  },
  {
    "_id": "5b0affc63e9b7f778c02c28c",
    "userName": "consequat",
    "password": "fugiat",
    "firstName": "Mclaughlin",
    "lastName": "Dale",
    "gradeId": "5b0116f896b61b3a5c135a1d"
  },
  {
    "_id": "5b0affc685e7165a20432425",
    "userName": "aliquip",
    "password": "et",
    "firstName": "Elvia",
    "lastName": "Rutledge",
    "gradeId": "5b0116f814169898640db32f"
  },
  {
    "_id": "5b0affc6cbefe48e056987c1",
    "userName": "duis",
    "password": "do",
    "firstName": "Sparks",
    "lastName": "Britt",
    "gradeId": "5b0116f814169898640db32f"
  },
  {
    "_id": "5b0affc6ddcb909292889046",
    "userName": "ut",
    "password": "quis",
    "firstName": "Sue",
    "lastName": "Hardin",
    "gradeId": "5b0116f814169898640db32f"
  },
  {
    "_id": "5b0affc6c7e1620e7bf4d10d",
    "userName": "aute",
    "password": "veniam",
    "firstName": "Frankie",
    "lastName": "Gonzales",
    "gradeId": "5b0116f814169898640db32f"
  }
]
