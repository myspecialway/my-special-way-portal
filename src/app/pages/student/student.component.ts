import { first } from 'rxjs/operators';
import { Component, OnInit, ViewChild, ElementRef } from '@angular/core';
import { MatTableDataSource, MatSort, MatDialog } from '@angular/material';
import { DeleteStudentDialogComponent } from './dialogs/delete/delete-student.dialog';
import { StudentService } from './services/student.service';
import Student from '../../models/student.model';
import { SubscriptionCleaner } from '../../decorators/SubscriptionCleaner.decorator';
import { Apollo } from 'apollo-angular';
import { UserProfileStateModel } from '../../apollo/state/resolvers/state.resolver';
import { GET_USER_PROFILE } from '../../apollo/state/queries/get-user-profile.query';
import { UserType } from '../../models/user.model';

@Component({
  selector: 'app-student',
  templateUrl: './student.component.html',
  styleUrls: ['./student.component.scss'],
})
export class StudentComponent implements OnInit {
  displayedColumns = ['studentName', 'gradeId', 'username', 'deleteUser'];
  dataSource = new MatTableDataSource<Student>();
  mayAddStudent = false;
  mayDeleteStudent = false;

  @ViewChild(MatSort)
  sort: MatSort;
  @ViewChild('table')
  table: ElementRef;

  @SubscriptionCleaner()
  subCollector;

  constructor(private studentService: StudentService, private dialog: MatDialog, private apollo: Apollo) {}

  ngOnInit() {
    this.dataSource.sort = this.sort;

    this.subCollector.add(
      this.studentService.getAllStudents().subscribe((data) => {
        this.dataSource.data = [...data];
      }),
    );

    // todo: create a currentUser / permissions service / directive to handle permissions.
    this.apollo
      .watchQuery<{ userProfile: UserProfileStateModel }>({
        query: GET_USER_PROFILE,
      })
      .valueChanges.subscribe((userProf) => {
        const currentType = userProf.data.userProfile.role;
        this.mayAddStudent = UserType[currentType] === UserType.PRINCIPLE;
        this.mayDeleteStudent = UserType[currentType] === UserType.PRINCIPLE;
      });
  }

  async deleteStudent(id: number, firstName: string, lastName: string, gradeId: string) {
    const dialogRef = this.dialog.open(DeleteStudentDialogComponent, {
      data: { id, firstName, lastName, gradeId },
    });

    dialogRef
      .afterClosed()
      .pipe(first())
      .subscribe(async (deletionConfirmed) => {
        if (!deletionConfirmed) {
          return;
        }

        try {
          await this.studentService.delete(id);
        } catch (error) {
          // TODO: implement error handling on UI
          console.error('Error handling not implemented');
          throw error;
        }
      });
  }
}
