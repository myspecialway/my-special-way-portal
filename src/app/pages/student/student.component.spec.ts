import { TestBed } from '@angular/core/testing';
import { NO_ERRORS_SCHEMA } from '@angular/core';
import {
  MatHeaderRowDef,
  MatRowDef,
  MatHeaderRow,
  MatDialog,
  MatSort,
  MatPaginator,
  MatPaginatorIntl,
} from '@angular/material';
import { StudentComponent } from './student.component';
import { StudentService } from './services/student.service';
import {
  Overlay,
  ScrollStrategyOptions,
  ScrollDispatcher,
  ViewportRuler,
  OverlayContainer,
  OverlayPositionBuilder,
  OverlayKeyboardDispatcher,
} from '@angular/cdk/overlay';
import { Platform } from '@angular/cdk/platform';
import { studentsTestData } from '../../../mocks/assets/students.mock';
import { Observable, Subject } from 'rxjs-compat';
import { Apollo } from 'apollo-angular';

describe('student component', () => {
  let studentServiceMock: Partial<StudentService>;
  let studentDialogMock: Partial<MatDialog>;
  let afterClosedMockFn: jest.Mock;
  const watchQueryMockedObservable: Subject<any>;

  beforeEach(async () => {
    studentServiceMock = {
      getAllStudents: jest.fn().mockReturnValueOnce(Observable.of(studentsTestData.students)),
      delete: jest.fn(),
      create: jest.fn(),
      update: jest.fn(),
    };

    afterClosedMockFn = jest.fn().mockReturnValue(Observable.of(true));
    studentDialogMock = {
      open: jest.fn().mockReturnValue({
        afterClosed: afterClosedMockFn,
      }),
    };

    const apolloMock = {
      watchQuery: () => ({
        valueChanges: watchQueryMockedObservable,
      }),
    };

    TestBed.configureTestingModule({
      imports: [],
      declarations: [StudentComponent, MatHeaderRow, MatRowDef, MatHeaderRowDef, MatSort, MatPaginator],
      providers: [
        StudentService,
        { provide: MatDialog, useValue: studentDialogMock },
        { provide: StudentService, useValue: studentServiceMock },
        Overlay,
        ScrollStrategyOptions,
        ScrollDispatcher,
        Platform,
        ViewportRuler,
        OverlayContainer,
        OverlayPositionBuilder,
        OverlayKeyboardDispatcher,
        MatPaginatorIntl,
        { provide: Apollo, useValue: apolloMock },
      ],
      schemas: [NO_ERRORS_SCHEMA],
    });
  });

  it('should render component as described in snapshot', () => {
    const fixture = TestBed.createComponent(StudentComponent);
    expect(fixture).toMatchSnapshot();
  });

  it('should open dialog when calling deleteStudent function', () => {
    const fixture = TestBed.createComponent(StudentComponent);
    fixture.componentInstance.deleteStudent(123, 'sad', 'asd', 'אשוח');
    const DialogMock = TestBed.get(MatDialog);
    expect(DialogMock.open).toHaveBeenCalled();
  });

  /* 
  ** temporry remove this as it fails with undefined of valueChanges
  * 
  it('should load students from service on page load ', () => {
    const fixture = TestBed.createComponent(StudentComponent);
    fixture.detectChanges();
    expect(studentServiceMock.getAllStudents).toHaveBeenCalled();
  });

  it('should load correct number of students ', async () => {
    const fixture = TestBed.createComponent(StudentComponent);
    fixture.detectChanges();
    await fixture.whenRenderingDone();

    expect(fixture.componentInstance.dataSource.data.length).toEqual(4);
  });
  ***/

  it('should load zero students in case of promise reject', async () => {
    const getAllStudentsMock = studentServiceMock.getAllStudents as jest.Mock;
    getAllStudentsMock.mockReset();
    getAllStudentsMock.mockRejectedValueOnce(0);

    const fixture = TestBed.createComponent(StudentComponent);
    expect(fixture.componentInstance.dataSource.data.length).toEqual(0);
  });

  it('component delete should call service delete', async () => {
    (studentServiceMock.delete as jest.Mock).mockImplementationOnce(() => {
      return Promise.resolve(1);
    });

    const fixture = TestBed.createComponent(StudentComponent);
    fixture.componentInstance.deleteStudent(1, 'name', 'name', 'asd');

    const serviceMock = TestBed.get(StudentService);
    expect(serviceMock.delete).toHaveBeenCalled();
  });

  it(`should not do anything if user hasn't confirmed delete dialog`, () => {
    // given
    expect.hasAssertions();
    afterClosedMockFn.mockReset();
    afterClosedMockFn.mockReturnValueOnce(Observable.of(false));

    // when
    const fixture = TestBed.createComponent(StudentComponent);
    fixture.componentInstance.deleteStudent(1, 'name', 'name', 'asd');

    // then
    expect(studentServiceMock.delete).not.toHaveBeenCalled();
  });

  it('should pipe promise rejection when deleteStudent fails', async () => {
    const deleteStudentsMock = studentServiceMock.delete as jest.Mock;
    deleteStudentsMock.mockReset();
    deleteStudentsMock.mockReturnValueOnce(Promise.reject('oh no!'));

    const fixture = TestBed.createComponent(StudentComponent).componentInstance;

    expect(fixture.deleteStudent(1, 'a', 'a', 'a')).rejects.toEqual('oh no!');
  });

  xit('should display an error to the user when deleteStudent fails', () => {});
});
