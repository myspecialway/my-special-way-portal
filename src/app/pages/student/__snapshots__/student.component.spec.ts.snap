// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`student component should render component as described in snapshot 1`] = `
<app-student
  apollo={[Function Object]}
  dataSource={[Function MatTableDataSource]}
  dialog={[Function Object]}
  displayedColumns={[Function Array]}
  mayAddStudent="false"
  mayDeleteStudent="false"
  sort={[Function MatSort]}
  studentService={[Function Object]}
  table="undefined"
>
  <div
    class="main-content msw-student-container"
    data-test-id="students-page"
  >
    <div
      class="container-fluid"
    >
      <div
        class="header-menu-container msw-table-tools-header"
      >
        <div
          class="header-table-count"
        >
          <div>
            <span
              class="header-table-count-label"
            >
              תלמידים
            </span>
            <span
              class="header-table-count-number"
            >
              (
            </span>
          </div>
        </div>
        <div
          class="example-header msw-table-tools-header"
        >
          
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="col-lg-12 col-md-12 col-sm-12"
        >
          <mat-table
            class="msw-table"
            matsort=""
            matsortdisableclear=""
          >
            
            
            
            
            
            
            
            
            
            
            --&gt; 
            
            
            
            
          </mat-table>
        </div>
      </div>
    </div>
  </div>
</app-student>
`;
